File: advent_of_code/day8/part2/main.go
package part2

import (
	"advent-of-code/advent_of_code/day8/part1"
	"advent-of-code/advent_of_code/helpers"
	"strings"
)

func Change(rs []string, indexToChange int) []string {
	switch rs[indexToChange][:3] {
	case "nop":
		rs[indexToChange] = strings.ReplaceAll(rs[indexToChange], "nop", "jmp")
	case "jmp":
		rs[indexToChange] = strings.ReplaceAll(rs[indexToChange], "jmp", "nop")
	}
	return rs
}

func Main(filepath string) int {

	rows, err := helpers.ReadFileLineByLine(filepath)
	if err != nil {
		panic(err)
	}
	indexes := make(map[int]bool)
	ind := 0
	acc := 0

	// make map of changable keys
	changableIndexes := make([]int, 0)
	for k, v := range rows {
		if v[:3] == "nop" || v[:3] == "jmp" {
			changableIndexes = append(changableIndexes, k)
		}
	}
	nextIndex := -1
	// main run
	for {

		newIndex, newAcc := part1.Action(rows[ind])
		ind += newIndex

		// swap condition
		if _, ok := indexes[ind]; ok {

			nextIndex += 1
			orig_rows, _ := helpers.ReadFileLineByLine(filepath)
			rows = Change(orig_rows, changableIndexes[nextIndex])
			ind = 0
			acc = 0
			indexes = make(map[int]bool)
			continue
		}

		// terminated successfully
		if ind >= len(rows) {
			acc += newAcc
			break
		}

		indexes[ind] = true
		acc += newAcc

	}

	return acc
}

File: advent_of_code/day8/part2/main.py
from typing import List, Iterable

from advent_of_code.day8.part1.main import action
from advent_of_code.helpers.helpers import read_file_line_by_line


def change(rs: List[str], index_to_change: int) -> List[str]:
    rsn: List[str] = rs
    if rs[index_to_change][:3] == "nop":
        rsn[index_to_change] = rs[index_to_change].replace("nop", "jmp")
    else:
        rsn[index_to_change] = rs[index_to_change].replace("jmp", "nop")

    return rsn


def main(filepath: str) -> int:

    rows: List[str] = read_file_line_by_line(filepath=filepath)
    orig_rows: List[str] = read_file_line_by_line(filepath=filepath)
    changable_indexes: Iterable[int] = iter(
        [i for i in range(len(rows)) if rows[i].split()[0] in ["nop", "jmp"]]
    )
    indexes: List[int] = []; ind: int = 0; acc: int = 0

    while True:
        new_index, new_acc = action(row=rows[ind])
        ind += new_index

        # swap condition
        if ind in indexes:
            switched_index: int = next(changable_indexes)
            indexes = []; ind = 0; acc = 0
            rows = orig_rows
            rows = change(rs=rows, index_to_change=switched_index)
            # there's some random memory leak here
            orig_rows = read_file_line_by_line(filepath=filepath)  
            assert orig_rows == read_file_line_by_line(filepath=filepath)
            continue

        # terminated successfully
        elif ind >= len(rows):
            acc += new_acc
            break
        
        else:
            indexes.append(ind)
            acc += new_acc

    return acc